### PR TITLE
TRUNK - 5068 Add tests to ProgramWorkFlowService

### PR DESCRIPTION
**Pull request by dilekamadushan**
https://issues.openmrs.org/browse/TRUNK-5068?filter=-1

## Description of what I changed

TRUNK - 5068 Add tests to ProgramWorkFlowService, Added test case to check whether it null concept throws error

## Checklist: I completed these to help reviewers :)

- [ x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

- [ x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

- [ x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

- [ x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

- [ x] All new and existing **tests passed**.

- [ x] My pull request is **based on the latest changes** of the master branch.

### CODE REVIEW

**Comment by dilekamadushan** [comment]
@teleivo Ok I will add more tests and commit

**Comment by dilekamadushan** [comment]
@teleivo I added some test cases to cover some of the branches. Looking forward for your feedback. Thanks.

File: api/src/test/java/org/openmrs/api/ProgramWorkflowServiceTest.java [hunk 1]
```diff
@@ -245,6 +245,18 @@ public void getProgramByUuid_shouldReturnNullIfNoObjectFoundWithGivenUuid() {
 		Assert.assertNull(Context.getProgramWorkflowService().getProgramByUuid("some invalid uuid"));
 	}
 	
+	/**
+	 * @see ProgramWorkflowService#getProgramByName(String)
+	 */
+	@Test
+	public void getProgramme_shouldGetProgrammeByName() {
```
Comment (teleivo) [inline api/src/test/java/org/openmrs/api/ProgramWorkflowServiceTest.java]: have you looked at the test coverage?

`getProgramByName` is fully tested, so please remove this test.

see current coverage of the class you are testing
https://coveralls.io/builds/14906713/source?filename=api%2Fsrc%2Fmain%2Fjava%2Forg%2Fopenmrs%2Fapi%2Fimpl%2FProgramWorkflowServiceImpl.java#L132

as the issue you are working on describes you should
* add unit tests (which require no database) into `ProgramWorkflowServiceUnitTest`
* tests that require a database into `ProgramWorkflowServiceTest`

we should try to add more tests without using a DB (we use an in-memory DB called h2) especially if they are so trivial like if I pass in null I expect such and such to happen. Please take the time to understand how the few tests in `ProgramWorkflowServiceUnitTest` work. We already have tons of tests which almost all run using a DB which in total takes quite some time. If we manage to write more tests without we will be able to reduce the time our builds run without loosing the security net that tests provide. Tests with DB are valuable but we should only have a selected few per service I think.

Comment (dilekamadushan) [reply api/src/test/java/org/openmrs/api/ProgramWorkflowServiceTest.java]: @The getProgramByName should be changed to getProgram. Thanks

Comment (dilekamadushan) [reply api/src/test/java/org/openmrs/api/ProgramWorkflowServiceTest.java]: @teleivo Shall I go on and create a new Exception class to handle concept being null ?

**Comment by dilekamadushan** [comment]
@teleivo I pushed my latest changes on the issue. Feel free to review. Thanks

**Comment by dkayiwa** [comment]
@dilekamadushan the ticket requires tests in two classes. This is just one of them.

**Comment by dilekamadushan** [comment]
@dkayiwa Yes It is true, but @teleivo suggested me to write one test method at a time

File: api/src/test/java/org/openmrs/api/ProgramWorkflowServiceUnitTest.java [hunk 2]
```diff
@@ -67,4 +72,15 @@ public void getProgramByName_shouldFailWhenTwoProgramsFoundWithSameName() {
...
 		pws.getProgramByName("A name");
 	}
+	
+	@Test
+	public void saveProgramme_shouldThrowErrorIfConceptOfProgrammeIsNull() {
+		
+		thrown.expect(APIException.class);
+		thrown.expectMessage("Program concept is required");
+		
+		Program program1 = new Program( 1 );
```
Comment (teleivo) [inline api/src/test/java/org/openmrs/api/ProgramWorkflowServiceUnitTest.java]: @dilekamadushan please dont write `new Program( 1 );` instead `new Program(1);` we do not put spaces around our parameters. Same below when calling `saveProgram` Fix your IDE style configuration if your IDE is doing that automatically https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE

File: api/src/test/java/org/openmrs/api/ProgramWorkflowServiceUnitTest.java [hunk 3]
```diff
@@ -67,4 +72,15 @@ public void getProgramByName_shouldFailWhenTwoProgramsFoundWithSameName() {
 		pws.setProgramWorkflowDAO(mockDao);
 		pws.getProgramByName("A name");
 	}
+	
+	@Test
+	public void saveProgramme_shouldThrowErrorIfConceptOfProgrammeIsNull() {
```
Comment (teleivo) [inline api/src/test/java/org/openmrs/api/ProgramWorkflowServiceUnitTest.java]: rename `saveProgramme_shouldThrowErrorIfConceptOfProgrammeIsNull` to `saveProgram_shouldFailIfProgramConceptIsNull`

File: api/src/test/java/org/openmrs/api/ProgramWorkflowServiceUnitTest.java [hunk 4]
```diff
@@ -30,6 +32,9 @@
 	
 	private ProgramWorkflowService pws;
 	
+	@Rule
+	public final ExpectedException thrown = ExpectedException.none();
```
Comment (teleivo) [inline api/src/test/java/org/openmrs/api/ProgramWorkflowServiceUnitTest.java]: rename to `exception`

**Comment by dilekamadushan** [comment]
@teleivo I modified the code on following your reviews. Thank you very much

**Comment by teleivo** [comment]
@dilekamadushan perfect! Can you now please add tests for the 2 missing branches of the `saveProgram` method you just tested right underneath the test you added. Just add them to this commit in this PR.

Have you tried creating the test coverage locally? So you can see the uncovered branches.
Either in your IDE or via `mvn clean package && mvn jacoco:report` which will spit out reports in I believe `target/site/jacoco/index.html`

**Comment by dilekamadushan** [comment]
@teleivo I added new test cases. Feel free to reviews. Thanks for your guidance.

**Comment by teleivo** [comment]
@dilekamadushan excellent job :)
were you able to generate coverage locally?

very similar tests are needed for
* ensureProgramIsSet
* ensureProgramWorkflowIsSet
* purgeProgram
* savePatientProgram
* purgePatientProgram

can you work on these?
after these we can work on a few other tests for this class where you could see a few other testing concepts.

thanks a lot!

**Comment by dilekamadushan** [comment]
@teleivo I added these test cases on my PR https://github.com/openmrs/openmrs-core/pull/2565. Feel free to review. Thanks
